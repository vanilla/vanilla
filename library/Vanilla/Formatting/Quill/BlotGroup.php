<?php
/**
 * @author Adam (charrondev) Charron <adam.c@vanillaforums.com>
 * @copyright 2009-2019 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Formatting\Quill;

use Vanilla\EmbeddedContent\Embeds\QuoteEmbed;
use Vanilla\Formatting\Quill\Blots\AbstractBlot;
use Vanilla\Formatting\Quill\Blots\Embeds\ExternalBlot;
use Vanilla\Formatting\Quill\Blots\Lines\AbstractLineTerminatorBlot;
use Vanilla\Formatting\Quill\Blots\Lines\CodeLineTerminatorBlot;
use Vanilla\Formatting\Quill\Blots\Lines\ListLineTerminatorBlot;
use Vanilla\Formatting\Quill\Blots\Lines\ParagraphLineTerminatorBlot;
use Vanilla\Formatting\Quill\Blots\Lines\SpoilerLineTerminatorBlot;
use Vanilla\Formatting\Quill\Blots\TextBlot;

/**
 * Class to represent a group of a quill blots. One group can contain:
 *
 * - Multiple Text formats in a single paragraph (as well as inline embeds).
 * - Multiple line blots if it is a line type grouping
 */
class BlotGroup {
    /** @var AbstractBlot[] */
    private $blots = [];

    /**
     * @var AbstractBlot[]
     *
     * Blots that can determine the surrounding tag over the other blot types.
     */
    private $overridingBlots = [
        CodeLineTerminatorBlot::class,
        AbstractLineTerminatorBlot::class,
    ];

    /**
     * Determine if this group is made up only of a break.
     *
     * @return bool
     */
    public function isBreakOnlyGroup(): bool {
        if (count($this->blots) !== 1) {
            return false;
        }

        $blot = $this->blots[0];

        return get_class($blot) === TextBlot::class && $this->blots[0]->getContent() === "";
    }

    /**
     * Check if the group's last matches the passed Blot class.
     *
     * @param string $blotClass - The class constructor of a Blot.
     * @param bool $needsExactMatch - If true, the class must match exactly, no child classes are allowed.
     *
     * @return bool
     */
    public function endsWithBlotOfType(string $blotClass, bool $needsExactMatch = false): bool {
        if (count($this->blots) === 0) {
            return false;
        }

        $lastBlot = $this->blots[count($this->blots) - 1];

        if ($needsExactMatch) {
            return get_class($lastBlot) === $blotClass;
        }

        return $lastBlot instanceof $blotClass;
    }

    /**
     * Determine if the group is empty.
     *
     * @return bool
     */
    public function isEmpty(): bool {
        return count($this->blots) === 0;
    }

    /** @var BlotGroup[] */
    private $nestedGroups = [];

    /**
     * Determine if one group can nest another one inside of it.
     *
     * @param BlotGroup $otherGroup
     * @return bool
     */
    public function canNest(BlotGroup $otherGroup): bool {
        $otherMainBlot = $otherGroup->getMainBlot();
        $ownMainBlot = $this->getMainBlot();
        if ($otherMainBlot instanceof ListLineTerminatorBlot
            && $ownMainBlot instanceof ListLineTerminatorBlot
            && $otherGroup->getNestingDepth() > $this->getNestingDepth()
        ) {
            return true;
        } else {
            return false;
        }
    }

    /**
     * @return int
     */
    public function getNestingDepth(): int {
        $ownMainBlot = $this->getMainBlot();
        return $ownMainBlot->getNestingDepth();
    }

    /**
     * Nest a blot group inside of this one.
     *
     * @param BlotGroup $blotGroup
     */
    public function nestGroup(BlotGroup $blotGroup) {
        $lastNestedGroup = $this->nestedGroups[count($this->nestedGroups) - 1] ?? null;
        if ($lastNestedGroup && $lastNestedGroup->canNest($blotGroup)) {
            $lastNestedGroup->nestGroup($blotGroup);
        } else {
            $this->nestedGroups[] = $blotGroup;
        }
    }

    /**
     * Render the block.
     *
     * @return string
     */
    public function render(): string {
        // We need a special exception for empty paragraph only blots.
        // This because the ParagraphLineTerminatorBlot does not render a break if there is other group content
        // But does render a break if it is alone. This is mostly to make styling easier.
        // We do not want to need make the paragraph terminator aware of its group.
        if (
            count($this->blots) === 1 &&
            $this->blots[0] instanceof ParagraphLineTerminatorBlot &&
            $this->blots[0]->getContent() === "\n"
        ) {
            return "<p><br></p>";
        }

        // Don't render empty groups.
        $result = '';
        $result .= $this->renderOpeningTag();
        $result .= $this->renderContent();
        $result .= $this->renderClosingTag();
        return $result;
    }

    /**
     * Render the content of a blot group.
     *
     * @return string
     */
    public function renderContent(): string {
        $result = '';
        $surroundTagBlot = $this->getMainBlot();

        if ($surroundTagBlot instanceof AbstractLineTerminatorBlot) {
            // Line blots have special rendering.
            $result .= $this->renderLineGroup();
        } else {
            foreach ($this->blots as $blot) {
                $result .= $blot->render();
            }
        }

        return $result;
    }

    /**
     * @return string
     */
    public function renderOpeningTag(): string {
        $surroundTagBlot = $this->getMainBlot();
        return $surroundTagBlot->getGroupOpeningTag();
    }

    /**
     * @return string
     */
    public function renderClosingTag(): string {
        $surroundTagBlot = $this->getMainBlot();
        return $surroundTagBlot->getGroupClosingTag();
    }

    /**
     * Render any nested groups inside of this one.
     *
     * If there are multiple nested groups they will share the start/end tag of the first nested group.
     *
     * @return string
     */
    private function renderNestedGroups(): string {
        $firstNestedGroup = $this->nestedGroups[0] ?? null;
        if (!$firstNestedGroup) {
            return "";
        }

        $result = "";
        $result .= $firstNestedGroup->renderOpeningTag();

        foreach ($this->nestedGroups as $nestedGroup) {
            // Only the first group will be used for the group tags.
            $result .= $nestedGroup->renderContent();
        }
        $result .= $firstNestedGroup->renderClosingTag();
        return $result;
    }

    /**
     * Render out a group with line terminators.
     *
     * We need to render potentially multiple inline blots than a line terminator, but the line terminator also
     * is responsible for the opening tag of the line and each line can have its special attributes even if
     * they are the same blot (eg. list indentation).
     *
     * @return string
     */
    public function renderLineGroup(): string {
        $result = "";

        // Look ahead and get all of the line terminators.
        $terminatorIndex = 0;
        $terminators = $this->getLineTerminators();
        $terminator = $terminators[$terminatorIndex];

        // Grab the first line terminator and start the line.
        $result .= $terminator->renderLineStart();

        foreach ($this->blots as $index => $blot) {
            $isLast = $index === count($this->blots) - 1;
            if ($blot instanceof AbstractLineTerminatorBlot) {
                // Render out the content of the line terminator (maybe nothing, maybe extra newlines).
                $result .= $terminator->render();
                if ($isLast) {
                    // render the nested groups inside of the last line, before it's closing tag. Eg.
                    // <li>Line 1 <ul>
                    //     <li>Line 1.1</li>
                    // </ul></li>
                    $result .= $this->renderNestedGroups();
                }
                // End the line.
                $result .= $terminator->renderLineEnd();

                // Start the next line.
                $terminatorIndex++;
                $terminator = $terminators[$terminatorIndex] ?? null;
                if ($terminator !== null) {
                    $result .= $terminator->renderLineStart();
                }
            } else {
                // Render out inline blots.
                $result .= $blot->render();
            }
        }

        return $result;
    }

    /**
     * Get all of the line terminators in the group.
     *
     * @return array
     */
    private function getLineTerminators(): array {
        $terminators = [];
        foreach ($this->blots as $blot) {
            if ($blot instanceof AbstractLineTerminatorBlot) {
                $terminators[] = $blot;
            }
        }
        return $terminators;
    }

    /**
     * Get all of the usernames that are mentioned in the blot group.
     *
     * Mentions that are inside of Blockquote's are excluded. We don't want to be sending notifications when big quote
     * replies build up.
     *
     * @return string[]
     */
    public function getMentionUsernames() {
        if ($this->getMainBlot() instanceof Blots\Lines\BlockquoteLineTerminatorBlot) {
            return [];
        }

        $names = [];
        foreach ($this->blots as $blot) {
            if ($blot instanceof Blots\Embeds\MentionBlot) {
                $names[] = $blot->getUsername();
            } elseif ($blot instanceof ExternalBlot) {
                $embed = $blot->getEmbed();
                if ($embed instanceof QuoteEmbed) {
                    $names[] = $embed->getUsername();
                }
            }
        }

        // De-duplicate the usernames.
        $names = array_unique($names);

        return $names;
    }

    /**
     * Add a blot to this block.
     *
     * @param AbstractBlot[] $blots
     */
    public function pushBlots(array $blots) {
        foreach ($blots as $blot) {
            $this->blots[] = $blot;
        }
    }

    /**
     * Get the position in the blots array of the first blot of the given type. Defaults to -1.
     *
     * @param string $blotClass The class string of the blot to check for.
     *
     * @return int
     */
    public function getIndexForBlotOfType(string $blotClass): int {
        $index = -1;

        foreach ($this->blots as $blotIndex => $blot) {
            if ($blot instanceof $blotClass) {
                $index = $blotIndex;
                break;
            }
        }

        return $index;
    }

    /**
     * Determine which blot should create the surrounding HTML tags of the group.
     *
     * @return AbstractBlot|null
     */
    public function getMainBlot(): ?AbstractBlot {
        if (count($this->blots) === 0) {
            return null;
        }
        $blot = $this->blots[0];
        $overridingBlot = $this->getOverrideBlot();

        return $overridingBlot ?? $blot;
    }

    /**
     * Get the primary blot for the group.
     *
     * @return null|AbstractBlot
     */
    public function getOverrideBlot(): ?AbstractBlot {
        foreach ($this->overridingBlots as $overridingBlot) {
            $index = $this->getIndexForBlotOfType($overridingBlot);
            if ($index >= 0) {
                return $this->blots[$index];
            }
        }

        return null;
    }

    /**
     * Simplify the blot group into an array of the classnames of its blots.
     *
     * @return string[]
     */
    public function getTestData(): array {
        $blots = [];
        foreach ($this->blots as $blot) {
            $blots[] = [
                "class" => get_class($blot),
                "content" => $blot->getContent(),
            ];
        }

        return $blots;
    }

    /**
     * Get the text value of a heading. This is not sanitized, so be sure to HTML escape before using.
     *
     * @return string
     */
    public function getUnsafeText(): string {
        $text = "";
        $mainBlot = $this->getMainBlot();
        if ($mainBlot instanceof SpoilerLineTerminatorBlot) {
            return \Gdn::translate("(Spoiler)") . "\n";
        }

        foreach ($this->blots as $blot) {
            $text .= $blot->getContent();
        }

        return $text;
    }
}
