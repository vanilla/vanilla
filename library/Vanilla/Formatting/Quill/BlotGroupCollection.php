<?php
/**
 * @author Adam (charrondev) Charron <adam.c@vanillaforums.com>
 * @copyright 2009-2019 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Formatting\Quill;

use Vanilla\Formatting\Formats\RichFormat;
use Vanilla\Formatting\FormatText;
use Vanilla\Formatting\Quill\Blots\AbstractBlot;
use Vanilla\Formatting\Quill\Blots\Embeds\ExternalBlot;
use Vanilla\Formatting\Quill\Blots\Lines\AbstractLineTerminatorBlot;
use Vanilla\Formatting\Quill\Nesting\NestingParentInterface;
use Vanilla\Formatting\Quill\Nesting\NestingParentRendererInterface;
use Vanilla\Formatting\TextDOMInterface;
use Vanilla\Formatting\TextFragmentInterface;
use Vanilla\Formatting\TextFragmentType;

/**
 * Class for sorting operations into blots and groups.
 */
class BlotGroupCollection implements \IteratorAggregate, TextDOMInterface {

    /** @var array[] The operations to parse. */
    private $operations;

    /** @var string[] The blots that we are allowed to parse out. */
    private $allowedBlotClasses;

    /** @var string The parsing mode to be passed through to blots when they are created. */
    private $parseMode = Parser::PARSE_MODE_NORMAL;

    /** @var BlotGroup[] The array of groups being built up. */
    private $groups;

    /** @var AbstractBlot The blot being build up in the iteration */
    private $inProgressBlot;

    /** @var BlotGroup the group being built up in the iteration. */
    private $inProgressGroup;

    /** @var array[] the line being built up in the iteration. */
    private $inProgressLine;

    /** @var array The primary operation being iterated. */
    private $currentOp;

    /** @var array The next operation being iterated. */
    private $nextOp;

    /** @var array The previous operation being iterated. */
    private $prevOp;

    // ITERABLE IMPLEMENTATION

    /**
     * @inheritdoc
     */
    public function getIterator() {
        return new \ArrayIterator($this->groups);
    }

    /**
     * BlotGroupCollection constructor.
     *
     * @param array[] $operations The operations to generate groups for.
     * @param string[] $allowedBlotClasses The class names of the blots we are allowed to create in the groups.
     * @param string $parseMode The parsing mode to create the blots with.
     */
    public function __construct(array $operations, array $allowedBlotClasses, string $parseMode) {
        $this->operations = $operations;
        $this->allowedBlotClasses = $allowedBlotClasses;
        $this->parseMode = $parseMode;
        $this->createBlotGroups();
    }

    /**
     * Push the current line into the group and reset the line.
     */
    private function clearLine() {
        $this->inProgressGroup->pushBlots($this->inProgressLine);
        $this->inProgressLine = [];
    }

    /**
     * Get the previous blot group.
     *
     * @return BlotGroup|null
     */
    private function getPreviousBlotGroup(): ?BlotGroup {
        return $this->groups[count($this->groups) - 1] ?? null;
    }

    /**
     * Push the group into our groups array and start a new one.
     * Do not push an empty group.
     */
    private function clearBlotGroup() {
        if ($this->inProgressGroup->isEmpty()) {
            return;
        }

        $currentGroup = $this->inProgressGroup;
        $prevGroup = $this->getPreviousBlotGroup();

        if ($prevGroup && $prevGroup->canNest($currentGroup)) {
            $prevGroup->nestGroup($currentGroup);
        } elseif ($prevGroup && $prevGroup->canMerge($currentGroup)) {
            // Merge the current group into the previous one.
            $prevGroup->pushBlots($currentGroup->getBlotsAndGroups());
        } else {
            $this->groups[] = $this->inProgressGroup;
        }

        $this->inProgressGroup = new BlotGroup();
    }

    /**
     * Create Blots and their groups.
     */
    private function createBlotGroups() {
        $this->inProgressGroup = new BlotGroup();
        $this->inProgressLine = [];
        $this->groups = [];

        for ($i = 0; $i < count($this->operations); $i++) {
            $this->currentOp = $this->operations[$i];
            $this->prevOp = $this->operations[$i - 1] ?? [];
            $this->nextOp = $this->operations[$i + 1] ?? [];
            $this->inProgressBlot = $this->getCurrentBlot();

            // In event of break blots we want to clear the group if applicable and the skip to the next item.
            if ($this->currentOp === Parser::BREAK_OPERATION) {
                $this->clearLine();
                $this->clearBlotGroup();
                continue;
            }

            if (($this->inProgressBlot->shouldClearCurrentGroup($this->inProgressGroup))) {
                // Ask the blot if it should close the current group.
                $this->clearBlotGroup();
            }

            // Ask the blot if it should close the current group.
            if ($this->inProgressBlot instanceof AbstractLineTerminatorBlot) {
                // Clear the line because we we had line terminator.
                $this->clearLine();
            }

            // Push the blot into our current line.
            $this->inProgressLine[] = $this->inProgressBlot;

            // Clear the line because we have a line terminator.
            if ($this->inProgressBlot instanceof AbstractLineTerminatorBlot) {
                $this->clearLine();
            }

            // Some block type blots get a group (and line) all to themselves.
            if ($this->inProgressBlot->isOwnGroup()) {
                $this->clearLine();
                $this->clearBlotGroup();
            }
        }

        // Iteration is done so we need to clear the line then the group.
        $this->clearLine();
        if (!$this->inProgressGroup->isEmpty()) {
            $this->clearBlotGroup();
        }
    }

    /**
     * Get the matching blot for a sequence of operations. Returns the default if no match is found.

     * @return AbstractBlot
     */
    public function getCurrentBlot(): AbstractBlot {
        // Fallback to a TextBlot if possible. Otherwise we fallback to rendering nothing at all.
        $blotClass = Blots\TextBlot::matches($this->currentOp) ? Blots\TextBlot::class : Blots\NullBlot::class;
        foreach ($this->allowedBlotClasses as $blot) {
            // Find the matching blot type for the current, last, and next operation.
            if ($blot::matches($this->currentOp)) {
                $blotClass = $blot;
                break;
            }
        }

        return new $blotClass($this->currentOp, $this->prevOp, $this->nextOp, $this->parseMode);
    }

    /**
     * @inheritDoc
     */
    public function getFragments(): array {
        $result = [];
        foreach ($this->groups as $i => $group) {
            $this->getFragmentsBlotGroup($group, $result);
        }

        return $result;
    }

    /**
     * Extract the text fragments from a blot group.
     *
     * @param BlotGroup $group
     * @param array $result Working result array.
     */
    private function getFragmentsBlotGroup(BlotGroup $group, array &$result = []) {
        if ($group->getBlotsAndGroups() instanceof AbstractBlot) {
            $result[] = new BlotGroupTextFragment($group, $this);
        } else {
            $from = 0;
            foreach ($group->getBlotsAndGroups() as $j => $blot) {
                /** @var AbstractBlot $blot */
                if ($blot instanceof AbstractLineTerminatorBlot) {
                    $result[] = new BlotGroupTextFragment($group, $this, $from, $j);
                    $from = $j + 1;
                } elseif ($blot instanceof ExternalBlot) {
                    $result[] = $this->makeExternalBlotFragments($blot);
                    $from = $j + 1;
                }

                if ($blot instanceof NestingParentRendererInterface) {
                    /** @var BlotGroup[] $children */
                    $children = $blot->getNestedGroups();
                    foreach ($children as $child) {
                        $this->getFragmentsBlotGroup($child, $result);
                    }
                }
            }
            // This is a sanity check to make sure all of the blots have been properly captured as fragments.
            if ($from !== count($group->getBlotsAndGroups())) {
                trigger_error("The blot group was not properly parsed into fragments.", E_USER_NOTICE);
            }
        }
    }

    /**
     * Get the allowed blot classes.
     *
     * @return string[]
     */
    public function getAllowedBlotClasses(): array {
        return $this->allowedBlotClasses;
    }

    /**
     * Get the blot parse mode.
     *
     * @return string
     */
    public function getParseMode(): string {
        return $this->parseMode;
    }

    /**
     * @inheritDoc
     */
    public function renderHTML(): string {
        $renderer = new Renderer();
        $result = $renderer->render($this);
        return $result;
    }

    /**
     * @inheritDoc
     */
    public function stringify(): FormatText {
        $result = [];
        foreach ($this->groups as $group) {
            $result = array_merge($result, $group->getOperations());
        }
        return new FormatText(json_encode($result), RichFormat::FORMAT_KEY);
    }

    /**
     * Get the blot groups array.
     *
     * @return BlotGroup[]
     */
    public function getGroups(): array {
        return $this->groups;
    }

    /**
     * Create text fragments from an instance of ExternalBlot.
     *
     * @param ExternalBlot $blot
     */
    private function makeExternalBlotFragments(ExternalBlot $blot) {
        $result = new TextFragmentCollection();

        $validFragments = [
            "url" => TextFragmentType::URL,
            "name" => TextFragmentType::TEXT,
            "body" => TextFragmentType::TEXT,
        ];
        foreach ($validFragments as $field => $type) {
            $path = "insert.embed-external.data.{$field}";
            $value = $blot->getCurrentOperationField($path, null);
            if (is_string($value)) {
                $result[$field] = new BlotPointerTextFragment($blot, $path, $type);
            }
        }

        return $result;
    }
}
