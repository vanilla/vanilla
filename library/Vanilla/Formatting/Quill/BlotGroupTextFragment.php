<?php
/**
 * @author Todd Burry <todd@vanillaforums.com>
 * @copyright 2009-2021 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Formatting\Quill;

use Vanilla\Formatting\Quill\Blots\Lines\CodeLineTerminatorBlot;
use Vanilla\Formatting\TextFragmentInterface;
use Vanilla\Formatting\TextFragmentType;

/**
 * An adaptor for blot groups that allow updating the text in each blot.
 */
final class BlotGroupTextFragment implements TextFragmentInterface {
    /**
     * @var BlotGroup
     */
    private $blotGroup;
    /**
     * @var BlotGroupCollection
     */
    private $parent;

    /**
     * @var object
     */
    private $from;

    /**
     * @var object
     */
    private $to;

    /**
     * BlotGroupTextFragment constructor.
     *
     * @param BlotGroup $blotGroup
     * @param BlotGroupCollection $parent
     * @param int $from The starting ordinal in the group that the text represents.
     * @param ?int $to The ending ordinal in the group that the text represents.
     */
    public function __construct(BlotGroup $blotGroup, BlotGroupCollection $parent, int $from = 0, int $to = null) {
        $this->blotGroup = $blotGroup;
        $this->parent = $parent;

        $blots = $blotGroup->getBlotsAndGroups();
        if ($to === null) {
            $to = count($blots) - 1;
        }
        $this->from = $blots[$from];
        $this->to = $blots[$to];

        $this->fromIndex = $from;
        $this->toIndex = $to;
    }

    /**
     * @inheritDoc
     */
    public function getInnerContent(): string {
        $r = $this->blotGroup->renderPartialLineGroupContent($this->getFromIndex());
        return $r;
    }

    /**
     * {@inheritDoc}
     */
    public function setInnerContent(string $text) {
        $blots = HtmlToBlotsParser::parseInlineHtml($text, $this->parent);

        $from = $this->getFromIndex();
        $to = $this->getToIndex();
        $this->blotGroup->replace($blots, $from, $to - 1);
    }

    /**
     * @inheritDoc
     */
    public function getFragmentType(): string {
        if ($this->to instanceof CodeLineTerminatorBlot) {
            return TextFragmentType::CODE;
        }
        return TextFragmentType::HTML;
    }

    /**
     * @return false|int|string
     */
    private function getFromIndex() {
        return array_search($this->from, $this->blotGroup->getBlotsAndGroups(), true);
    }

    /**
     * @return false|int|string
     */
    private function getToIndex() {
        return array_search($this->to, $this->blotGroup->getBlotsAndGroups(), true);
    }
}
