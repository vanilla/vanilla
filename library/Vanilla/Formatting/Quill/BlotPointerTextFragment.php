<?php
/**
 * @copyright 2009-2021 Vanilla Forums Inc.
 * @license GPL-2.0-only
 */

namespace Vanilla\Formatting\Quill;

use Vanilla\Formatting\Quill\Blots\AbstractBlot;
use Vanilla\Formatting\TextFragmentInterface;
use Vanilla\Formatting\TextFragmentType;

/**
 * Class BlotPointerTextFragment
 */
class BlotPointerTextFragment implements TextFragmentInterface {

    /**
     * @var AbstractBlot
     */
    private $blot;

    /**
     * @var string
     */
    private $path;

    /**
     * @var string
     */
    private $type;


    /**
     * BlotPointerTextFragment constructor.
     *
     * @param AbstractBlot $blot
     * @param string $path
     * @param string $type
     */
    public function __construct(AbstractBlot $blot, string $path, string $type = TextFragmentType::TEXT) {
        $this->blot = $blot;
        $this->path = $path;
        $this->type = $type;
    }

    /**
     * @inheritDoc
     */
    public function getInnerContent(): string {
        return $this->blot->getCurrentOperationField($this->path);
    }

    /**
     * Set the text of the fragment.
     *
     * @param string $text
     */
    public function setInnerContent(string $text) {
        $this->blot->setCurrentOperationField($this->path, $text);
    }

    /**
     * @inheritDoc
     */
    public function getFragmentType(): string {
        return $this->type;
    }
}
