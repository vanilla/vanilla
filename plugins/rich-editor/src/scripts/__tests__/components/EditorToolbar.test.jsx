import React from "react";
import { mount, shallow } from "enzyme";
import EditorMenuItem from "../../components/EditorMenuItem";
import EditorToolbar from "../../components/EditorToolbar";
import Quill from "quill/core";

jest.mock('quill/core');

test("matches snapshot", () => {
    const quill = new Quill();
    const toolbar = shallow(
        <EditorToolbar quill={quill}/>
    );

    expect(toolbar).toMatchSnapshot();
});

test("generates correct number of <EditorMenuItem /> components", () => {
    const menuItems = {
        foo: {},
        bar: {},
        other: {},
        thing: {},
    };
    const quill = new Quill();
    const toolbar = shallow(
        <EditorToolbar quill={quill} menuItems={menuItems}/>
    );

    expect(toolbar.find(EditorMenuItem).length).toBe(4);
});

test("can receive a custom formatter for the menu item click handler.", () => {
    const mockFormatter = jest.fn();

    const menuItems = {
        bold: {
            active: false,
            formatter: mockFormatter,
        },
    };
    const quill = new Quill();
    const toolbar = mount(
        <EditorToolbar quill={quill} menuItems={menuItems}/>
    );

    toolbar.find(".richEditor-button").simulate("click");
    toolbar.find(".richEditor-button").simulate("click");

    expect(mockFormatter.mock.calls.length).toBe(2);
});

describe("update", () => {
    let toolbar;

    beforeAll(() => {
        const quill = new Quill();
        const menuItems = {
            bold: {
                active: false,
            },
        };

        toolbar = shallow(
            <EditorToolbar quill={quill} menuItems={menuItems}/>
        );
    });

    it("updates boolean type ", () => {

    });
});
