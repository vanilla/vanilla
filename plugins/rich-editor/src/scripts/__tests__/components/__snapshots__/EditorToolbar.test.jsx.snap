// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditorToolbar
    quill={
      Quill {
        "addContainer": [MockFunction],
        "blur": [MockFunction],
        "deleteText": [MockFunction],
        "disable": [MockFunction],
        "enable": [MockFunction],
        "focus": [MockFunction],
        "format": [MockFunction],
        "formatLine": [MockFunction],
        "formatText": [MockFunction],
        "getBounds": [MockFunction],
        "getContents": [MockFunction],
        "getFormat": [MockFunction],
        "getIndex": [MockFunction],
        "getLeaf": [MockFunction],
        "getLength": [MockFunction],
        "getLine": [MockFunction],
        "getLines": [MockFunction],
        "getModule": [MockFunction],
        "getSelection": [MockFunction],
        "getText": [MockFunction],
        "hasFocus": [MockFunction],
        "insertEmbed": [MockFunction],
        "insertText": [MockFunction],
        "isEnabled": [MockFunction],
        "off": [MockFunction],
        "on": [MockFunction] {
          "calls": Array [
            Array [
              "editor-change",
              [Function],
            ],
            Array [
              "scroll-optimize",
              [Function],
            ],
          ],
        },
        "once": [MockFunction],
        "pasteHTML": [MockFunction],
        "removeFormat": [MockFunction],
        "scrollIntoView": [MockFunction],
        "setContents": [MockFunction],
        "setSelection": [MockFunction],
        "setText": [MockFunction],
        "update": [MockFunction],
        "updateContents": [MockFunction],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "aria-label": "Inline Level Formatting Menu",
      "children": <ul
        aria-label="Inline Level Formatting Menu"
        className="richEditor-menuItems MenuItems"
        role="menubar"
      >
        <EditorMenuItem
          clickHandler={[Function]}
          isActive={false}
          propertyName="bold"
        />
        <EditorMenuItem
          clickHandler={[Function]}
          isActive={false}
          propertyName="italic"
        />
        <EditorMenuItem
          clickHandler={[Function]}
          isActive={false}
          propertyName="strike"
        />
        <EditorMenuItem
          clickHandler={[Function]}
          isActive={false}
          propertyName="code"
        />
        <EditorMenuItem
          clickHandler={[Function]}
          isActive={false}
          propertyName="link"
        />
      </ul>,
      "className": "richEditor-menu",
      "role": "dialog",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-label": "Inline Level Formatting Menu",
        "children": Array [
          <EditorMenuItem
            clickHandler={[Function]}
            isActive={false}
            propertyName="bold"
          />,
          <EditorMenuItem
            clickHandler={[Function]}
            isActive={false}
            propertyName="italic"
          />,
          <EditorMenuItem
            clickHandler={[Function]}
            isActive={false}
            propertyName="strike"
          />,
          <EditorMenuItem
            clickHandler={[Function]}
            isActive={false}
            propertyName="code"
          />,
          <EditorMenuItem
            clickHandler={[Function]}
            isActive={false}
            propertyName="link"
          />,
        ],
        "className": "richEditor-menuItems MenuItems",
        "role": "menubar",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "class",
          "props": Object {
            "clickHandler": [Function],
            "isActive": false,
            "propertyName": "bold",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "clickHandler": [Function],
            "isActive": false,
            "propertyName": "italic",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "clickHandler": [Function],
            "isActive": false,
            "propertyName": "strike",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "class",
          "props": Object {
            "clickHandler": [Function],
            "isActive": false,
            "propertyName": "code",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "4",
          "nodeType": "class",
          "props": Object {
            "clickHandler": [Function],
            "isActive": false,
            "propertyName": "link",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "ul",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-label": "Inline Level Formatting Menu",
        "children": <ul
          aria-label="Inline Level Formatting Menu"
          className="richEditor-menuItems MenuItems"
          role="menubar"
        >
          <EditorMenuItem
            clickHandler={[Function]}
            isActive={false}
            propertyName="bold"
          />
          <EditorMenuItem
            clickHandler={[Function]}
            isActive={false}
            propertyName="italic"
          />
          <EditorMenuItem
            clickHandler={[Function]}
            isActive={false}
            propertyName="strike"
          />
          <EditorMenuItem
            clickHandler={[Function]}
            isActive={false}
            propertyName="code"
          />
          <EditorMenuItem
            clickHandler={[Function]}
            isActive={false}
            propertyName="link"
          />
        </ul>,
        "className": "richEditor-menu",
        "role": "dialog",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-label": "Inline Level Formatting Menu",
          "children": Array [
            <EditorMenuItem
              clickHandler={[Function]}
              isActive={false}
              propertyName="bold"
            />,
            <EditorMenuItem
              clickHandler={[Function]}
              isActive={false}
              propertyName="italic"
            />,
            <EditorMenuItem
              clickHandler={[Function]}
              isActive={false}
              propertyName="strike"
            />,
            <EditorMenuItem
              clickHandler={[Function]}
              isActive={false}
              propertyName="code"
            />,
            <EditorMenuItem
              clickHandler={[Function]}
              isActive={false}
              propertyName="link"
            />,
          ],
          "className": "richEditor-menuItems MenuItems",
          "role": "menubar",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "class",
            "props": Object {
              "clickHandler": [Function],
              "isActive": false,
              "propertyName": "bold",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "class",
            "props": Object {
              "clickHandler": [Function],
              "isActive": false,
              "propertyName": "italic",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "class",
            "props": Object {
              "clickHandler": [Function],
              "isActive": false,
              "propertyName": "strike",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "3",
            "nodeType": "class",
            "props": Object {
              "clickHandler": [Function],
              "isActive": false,
              "propertyName": "code",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "4",
            "nodeType": "class",
            "props": Object {
              "clickHandler": [Function],
              "isActive": false,
              "propertyName": "link",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "ul",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
